File: packages/xyd-opensdk-dotnet/src/cstype.ts
import type { NamedType, TypeRef } from '@xyd-js/opensdk-core';

import { pascalCase } from './naming';

/**
 * Map an IR TypeRef to a C# type expression. Named refs resolve through the
 * symbol table: struct/enum refs become their PascalCased class/enum name;
 * an ALIAS resolves to its underlying C# type (C# has no namespace-level type
 * alias, so aliases don't get their own class); a UNION maps to `object`
 * (phase-2 seam — discriminated-union modeling is not attempted yet).
 * Scalars map to C# built-ins; arrays -> List<T>; maps -> Dictionary<string, T>.
 */
export function csType(ref: TypeRef | undefined, types: Map<string, NamedType>): string {
  if (!ref) return 'object';
  switch (ref.kind) {
    case 'scalar':
      return csScalar(ref.scalar, ref.format);
    case 'ref': {
      if (!ref.name) return 'object';
      const named = types.get(ref.name);
      if (named?.kind === 'alias') return csType(named.of as TypeRef | undefined, types);
      if (named?.kind === 'union') return 'object';
      return pascalCase(ref.name); // struct / enum (or unknown placeholder -> treated as class)
    }
    case 'array':
      return `List<${csType(ref.items as TypeRef | undefined, types)}>`;
    case 'map':
      return `Dictionary<string, ${csType(ref.values as TypeRef | undefined, types)}>`;
    default:
      return 'object';
  }
}

function csScalar(scalar: string | undefined, format: string | undefined): string {
  switch (scalar) {
    case 'string':
      return format === 'binary' ? 'byte[]' : 'string';
    case 'integer':
      return format === 'int32' ? 'int' : 'long';
    case 'number':
      return format === 'float' ? 'float' : 'double';
    case 'boolean':
      return 'bool';
    default:
      return 'object';
  }
}

/** Wrap a C# type in a nullable (`T?`) unless it already is one. */
export function nullable(type: string): string {
  return type.endsWith('?') ? type : `${type}?`;
}

/** Whether a TypeRef is a binary blob (string format=binary) — an upload/download payload. */
export function isBinaryRef(ref: TypeRef | undefined): boolean {
  return ref?.kind === 'scalar' && ref.scalar === 'string' && ref.format === 'binary';
}

/**
 * Whether a TypeRef ultimately carries binary bytes (`format: binary`), following
 * array items and named union/alias refs (struct/enum refs are never binary). A
 * `seen` set guards recursive unions. Mirrors the Python emitter's isBinaryTypeRef:
 * a json body carrying such a field is routed through the multipart encoder.
 */
export function isBinaryTypeRef(
  ref: TypeRef | undefined,
  types: Map<string, NamedType>,
  seen = new Set<string>(),
): boolean {
  if (!ref) return false;
  if (ref.kind === 'scalar') return ref.scalar === 'string' && ref.format === 'binary';
  if (ref.kind === 'array') return isBinaryTypeRef(ref.items as TypeRef | undefined, types, seen);
  if (ref.kind === 'ref' && ref.name) {
    if (seen.has(ref.name)) return false;
    seen.add(ref.name);
    const named = types.get(ref.name);
    if (!named) return false;
    if (named.kind === 'union') return (named.variants || []).some((v) => isBinaryTypeRef(v as TypeRef, types, seen));
    if (named.kind === 'alias') return isBinaryTypeRef(named.of as TypeRef | undefined, types, seen);
  }
  return false;
}

File: crates/xyd_opensdk_dotnet/src/cstype.rs
//! cstype.ts — IR TypeRef → C# type expression + binary-payload detection.

use std::collections::BTreeMap;

use serde_json::Value;

use crate::jsrt::pascal_case;

/// The symbol table: name → NamedType (as raw JSON Value).
pub type Types<'a> = &'a BTreeMap<String, Value>;

/// Map an IR TypeRef to a C# type expression.
pub fn cs_type(ref_: Option<&Value>, types: Types) -> String {
    let r = match ref_ {
        Some(r) => r,
        None => return "object".to_string(),
    };
    match r.get("kind").and_then(Value::as_str) {
        Some("scalar") => cs_scalar(
            r.get("scalar").and_then(Value::as_str),
            r.get("format").and_then(Value::as_str),
        ),
        Some("ref") => {
            let name = match r.get("name").and_then(Value::as_str) {
                Some(n) if !n.is_empty() => n,
                _ => return "object".to_string(),
            };
            if let Some(named) = types.get(name) {
                match named.get("kind").and_then(Value::as_str) {
                    Some("alias") => return cs_type(named.get("of"), types),
                    Some("union") => return "object".to_string(),
                    _ => {}
                }
            }
            pascal_case(name)
        }
        Some("array") => format!("List<{}>", cs_type(r.get("items"), types)),
        Some("map") => format!("Dictionary<string, {}>", cs_type(r.get("values"), types)),
        _ => "object".to_string(),
    }
}

fn cs_scalar(scalar: Option<&str>, format: Option<&str>) -> String {
    match scalar {
        Some("string") => {
            if format == Some("binary") {
                "byte[]".to_string()
            } else {
                "string".to_string()
            }
        }
        Some("integer") => {
            if format == Some("int32") {
                "int".to_string()
            } else {
                "long".to_string()
            }
        }
        Some("number") => {
            if format == Some("float") {
                "float".to_string()
            } else {
                "double".to_string()
            }
        }
        Some("boolean") => "bool".to_string(),
        _ => "object".to_string(),
    }
}

/// Wrap a C# type in a nullable (`T?`) unless it already is one.
pub fn nullable(t: &str) -> String {
    if t.ends_with('?') {
        t.to_string()
    } else {
        format!("{t}?")
    }
}

/// Whether a TypeRef ultimately carries binary bytes (`format: binary`),
/// following array items and named union/alias refs (`seen` guards recursion).
pub fn is_binary_type_ref(ref_: Option<&Value>, types: Types, seen: &mut Vec<String>) -> bool {
    let r = match ref_ {
        Some(r) => r,
        None => return false,
    };
    match r.get("kind").and_then(Value::as_str) {
        Some("scalar") => {
            r.get("scalar").and_then(Value::as_str) == Some("string")
                && r.get("format").and_then(Value::as_str) == Some("binary")
        }
        Some("array") => is_binary_type_ref(r.get("items"), types, seen),
        Some("ref") => {
            let name = match r.get("name").and_then(Value::as_str) {
                Some(n) if !n.is_empty() => n,
                _ => return false,
            };
            if seen.iter().any(|s| s == name) {
                return false;
            }
            seen.push(name.to_string());
            let named = match types.get(name) {
                Some(n) => n,
                None => return false,
            };
            match named.get("kind").and_then(Value::as_str) {
                Some("union") => named
                    .get("variants")
                    .and_then(Value::as_array)
                    .map(|vs| vs.iter().any(|v| is_binary_type_ref(Some(v), types, seen)))
                    .unwrap_or(false),
                Some("alias") => is_binary_type_ref(named.get("of"), types, seen),
                _ => false,
            }
        }
        _ => false,
    }
}
